plugin-locationcomponent/src/main/java/com/mapbox/maps/plugin/locationcomponent/LocationCompassEngine.kt

```diff
@@ -0,0 +1,321 @@
+  private val compassListeners = mutableListOf<CompassListener>()
+
+  // Not all devices have a compassSensor
+  private var compassSensor: Sensor? = null
+  private var gravitySensor: Sensor? = null
+  private var magneticFieldSensor: Sensor? = null
+
+  private val rotationMatrix = FloatArray(9)
+  private var rotationVectorValue: FloatArray? = null
+  private var compassUpdateNextTimestamp: Long = 0
+  private var gravityValues = FloatArray(3)
+  private var magneticValues = FloatArray(3)
+
+  init {
+    compassSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
+    if (compassSensor == null) {
+      Logger.w(
+        TAG,
+        "Rotation vector sensor not supported on device, falling back to accelerometer and magnetic field."
+      )
+      gravitySensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
+      magneticFieldSensor = sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)
+    }
+  }
+
+  fun addCompassListener(compassListener: CompassListener) {
+    if (compassListeners.isEmpty()) {
+      registerSensorListeners()
+    }
+    if (!compassListeners.contains(compassListener)) {
+      compassListeners.add(compassListener)
+    }
+  }
+
+  fun removeCompassListener(compassListener: CompassListener) {
+    compassListeners.remove(compassListener)
+    if (compassListeners.isEmpty()) {
+      unregisterSensorListeners()
+    }
+  }
+
+  override fun onSensorChanged(event: SensorEvent) {
+    when (event.sensor.type) {
+      Sensor.TYPE_ROTATION_VECTOR -> {
+        rotationVectorValue = event.values
+        updateOrientation()
+      }
+      Sensor.TYPE_ACCELEROMETER -> {
+        gravityValues = lowPassFilter(event.values, gravityValues)
+        updateOrientation()
+      }
+      Sensor.TYPE_MAGNETIC_FIELD -> {
+        magneticValues = lowPassFilter(event.values, magneticValues)
+        updateOrientation()
+      }
+    }
+  }
+
+  override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
+    // no need
+  }
+
+  private fun updateOrientation() {
+    // check when the last time the compass was updated, return if too soon.
+    val currentTime = SystemClock.elapsedRealtime()
+    if (currentTime < compassUpdateNextTimestamp) {
+      return
+    }
+    if (rotationVectorValue != null) {
+      SensorManager.getRotationMatrixFromVector(rotationMatrix, rotationVectorValue)
+    } else {
+      // Get rotation matrix given the gravity and geomagnetic matrices
+      SensorManager.getRotationMatrix(rotationMatrix, null, gravityValues, magneticValues)
+    }
+    val (worldAxisForDeviceAxisX, worldAxisForDeviceAxisY) = getWorldAxisFromRotation()
+    val adjustedRotationMatrix = FloatArray(9)
+    SensorManager.remapCoordinateSystem(
+      rotationMatrix, worldAxisForDeviceAxisX,
+      worldAxisForDeviceAxisY, adjustedRotationMatrix
+    )
+
+    // Transform rotation matrix into azimuth/pitch/roll
+    val orientation = FloatArray(3)
```

Comment:
`updateOrientation` could be called pretty often, let's get rid of allocating that array each time and move it to private val and re-use